Mathlib/RingTheory/Bialgebra.lean

```diff
@@ -0,0 +1,83 @@
+/-
+Copyright (c) 2024 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey
+-/
+import Mathlib.RingTheory.Coalgebra
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Bialgebras
+
+In this file we define `Bialgebra`, and provide instances for:
+
+* Commutative semirings: `CommSemiring.toBialgebra`
+
+## References
+
+* <https://en.wikipedia.org/wiki/Bialgebra>
+-/
+
+suppress_compilation
+
+universe u v
+
+open scoped TensorProduct
+
+/-- A bialgebra over a commutative (semi)ring `R` is both an algebra and a coalgebra over `R`, such
+that the counit and comultiplication are algebra morphisms. -/
+class Bialgebra (R : Type u) (A : Type v) [CommSemiring R] [Semiring A] extends
+    Algebra R A, Coalgebra R A where
+  -- The counit is an algebra morphism
+  /-- The counit on a bialgebra preserves 1. -/
+  counit_one : counit 1 = 1
+  /-- The counit on a bialgebra preserves multiplication. -/
+  counit_mul : ∀ a₁ a₂ : A, counit (a₁ * a₂) = counit a₁ * counit a₂
```

Comment:
I think it will turn out to be easier to make instances of this if you use
```suggestion
  mul_compr₂_counit : (LinearMap.mul R A).compr₂ counit = (LinearMap.mul R R).compl₁₂ counit counit
```
and then prove a `counit_mul` lemma using `LinearMap.ext₂`. Similarly for `comul`.

Reply:
I find this comment hard to believe, but more to the point I made this change on the tube home and now I think that this makes the definition super-ugly and incomprehensible to a passing mathematician. Thus I propose instead leaving the definition alone (i.e. reverting the last two commits) and insead making a constructor with the incomprehensible fields instead. Would you be happy with this @eric-wieser ? I don't immediately see the advantage of making the definition itself obscure.

Reply:
The arguments for changing the definition are:
* I claim that *every* `instance` of this typeclass will want to use my proposed constructor. Anyone who doesn't will be making more work for themselves, and if it's not the default constructor, these people may not even be aware that the "easy" constructor exists. Of course, docstrings can lessen this issue.
* Only the default constructor works with `where` notation
* The fields for `Coalgebra` are stated in this point-free way already (for the same reason), so this is in some sense consistent

I agree that this definition is unfriendly to a mathematician reading it, but so is `Group` and many other typeclasses in mathlib. I don't yet have a good answer to how to solve that problem. Maybe it's worth getting a second opinion from Zulip on this.

Reply:
OK after consulting with someone else (who thinks you're right) I have gone with the obscure definition as the default, and added the less scary one as `Bialgebra.mk'`. Hopefully the docstrings are helpful if people are confused.